cocotb/triggers.py

```diff
@@ -917,15 +944,24 @@ async def with_timeout(trigger, timeout_time, timeout_unit="step"):
 
     .. deprecated:: 1.5
         Using None as the the *timeout_unit* argument is deprecated, use ``'step'`` instead.
-   """
 
+    .. versionchanged:: 1.6
+        Support passing :term:`python:coroutine`\ s.
+    """
     if timeout_unit is None:
         warnings.warn(
             'Using timeout_unit=None is deprecated, use timeout_unit="step" instead.',
             DeprecationWarning, stacklevel=2)
-        timeout_unit="step"  # don't propagate deprecated value
+        timeout_unit = "step"  # don't propagate deprecated value
+    if inspect.iscoroutine(trigger):
+        trigger = cocotb.fork(trigger)
+        shielded = False
+    else:
+        shielded = True
     timeout_timer = cocotb.triggers.Timer(timeout_time, timeout_unit)
     res = await First(timeout_timer, trigger)
+    if isinstance(res, cocotb.decorators.RunningTask) and not shielded:
+        trigger.kill()
     if res is timeout_timer:
         raise cocotb.result.SimTimeoutError
```

Comment:
There are two problems with this:
1. `await Join(task)` gives the return value of the task, not a `Join` or `RunningTask` object.
2. even if `res` was the forked task, it would mean the timeout did _not_ occur and the task finished, so `kill()` is redundant.
```suggestion
    if res is timeout_timer:
        if not shielded:
            trigger.kill()
        raise cocotb.result.SimTimeoutError
```